wasm/jit/engine.go

```diff
@@ -0,0 +1,329 @@
+				case reflect.Uint32, reflect.Uint64:
+					val.SetUint(raw)
+				case reflect.Int32, reflect.Int64:
+					val.SetInt(int64(raw))
+				}
+				in[i] = val
+			}
+			val := reflect.New(tp.In(0)).Elem()
+			val.Set(reflect.ValueOf(ctx))
+			in[0] = val
+			for _, ret := range f.HostFunction.Call(in) {
+				switch ret.Kind() {
+				case reflect.Float64, reflect.Float32:
+					e.push(math.Float64bits(ret.Float()))
+				case reflect.Uint32, reflect.Uint64:
+					e.push(ret.Uint())
+				case reflect.Int32, reflect.Int64:
+					e.push(uint64(ret.Int()))
+				default:
+					panic("invalid return type")
+				}
+			}
+		}
+		e.hostFunctions[id] = hf
+	} else {
+		id := e.compiledWasmFunctionIndex[f]
+		if e.compiledWasmFunctions[id] != nil {
+			// Already compiled.
+			return nil
+		}
+		cf, err := e.compileWasmFunction(f)
+		if err != nil {
+			return fmt.Errorf("failed to compile Wasm function: %w", err)
+		}
+		e.compiledWasmFunctions[id] = cf
+	}
+	return nil
+}
+
+func NewEngine() wasm.Engine {
+	return newEngine()
+}
+
+const initialStackSize = 100
+
+func newEngine() *engine {
+	e := &engine{
+		stack:                     make([]uint64, initialStackSize),
+		compiledWasmFunctionIndex: make(map[*wasm.FunctionInstance]int64),
+		hostFunctionIndex:         make(map[*wasm.FunctionInstance]int64),
+	}
+	return e
+}
+
+func (e *engine) pop() (ret uint64) {
+	ret = e.stack[e.currentBaseStackPointer+e.currentStackPointer-1]
+	e.currentStackPointer--
+	return
+}
+
+func (e *engine) push(v uint64) {
+	e.stack[e.currentBaseStackPointer+e.currentStackPointer] = v
+	e.currentStackPointer++
+}
+
+type jitStatusCodes uint32
+
+const (
+	jitStatusReturned jitStatusCodes = iota
+	jitStatusCallWasmFunction
+	jitStatusCallBuiltInFunction
+	jitStatusCallHostFunction
+	// TODO: trap, etc?
+)
+
+var (
+	engineStackSliceOffset              = int64(unsafe.Offsetof((&engine{}).stack))
+	engineCurrentStackPointerOffset     = int64(unsafe.Offsetof((&engine{}).currentStackPointer))
+	engineCurrentBaseStackPointerOffset = int64(unsafe.Offsetof((&engine{}).currentBaseStackPointer))
+	engineJITStatusOffset               = int64(unsafe.Offsetof((&engine{}).jitCallStatusCode))
+	engineFunctionCallIndexOffset       = int64(unsafe.Offsetof((&engine{}).functionCallIndex))
+	engineContinuationAddressOffset     = int64(unsafe.Offsetof((&engine{}).continuationAddressOffset))
+)
+
+type callFrame struct {
+	continuationAddress      uintptr
+	continuationStackPointer uint64
+	baseStackPointer         uint64
+	f                        *compiledWasmFunction
+	caller                   *callFrame
+}
+
+func (c *callFrame) String() string {
+	return fmt.Sprintf(
+		"[continuation address=%d, continuation stack poitner=%d, base stack pointer=%d]",
+		c.continuationAddress, c.continuationStackPointer, c.baseStackPointer,
+	)
+}
+
+type compiledWasmFunction struct {
+	inputNum, outputNum uint64
+	codeSegment         []byte
+	memoryInst          *wasm.MemoryInstance
```

Comment:
rename to mem or memory? inst is repeating the type name and not important, right?

Reply:
a1b00ed done